store/file/q4.go

```diff
@@ -0,0 +1,183 @@
+
+// Q4 implements eds.Accessor as an FS file.
+// It stores the Q4 of the EDS and its metadata in file's header.
+// It currently implements access only to Q4
+// and can be extended to serve the whole EDS as the need comes.
+type Q4 struct {
+	hdr  *headerV0
+	file *os.File
+}
+
+// CreateQ4 creates a new file under given FS path and
+// writes the Q4 into it out of given EDS.
+// It ensures FS is synced after writing finishes.
+func CreateQ4(
+	path string,
+	roots *share.AxisRoots,
+	eds *rsmt2d.ExtendedDataSquare,
+) error {
+	mod := os.O_RDWR | os.O_CREATE | os.O_EXCL // ensure we fail if already exist
+	f, err := os.OpenFile(path, mod, filePermissions)
+	if err != nil {
+		return fmt.Errorf("creating Q4 file: %w", err)
+	}
+
+	hdr := &headerV0{
+		fileVersion: fileV0,
+		fileType:    q4,
+		shareSize:   share.Size,
+		squareSize:  uint16(eds.Width()),
+		datahash:    roots.Hash(),
+	}
+
+	if err := writeQ4File(f, eds, hdr); err != nil {
+		errCls := f.Close()
+		errRm := RemoveQ4(path)
+		return errors.Join(err, errCls, errRm)
+	}
+
+	if err = f.Close(); err != nil {
+		return fmt.Errorf("closing created Q4 file: %w", err)
+	}
+
+	return nil
+}
+
+// writeQ4File full Q4 content into OS File.
+func writeQ4File(f *os.File, eds *rsmt2d.ExtendedDataSquare, hdr *headerV0) error {
+	// buffering gives us ~4x speed up
+	buf := bufio.NewWriterSize(f, writeBufferSize)
+
+	if err := writeHeader(buf, hdr); err != nil {
+		return fmt.Errorf("writing Q4 header: %w", err)
+	}
+
+	if err := writeQ4(buf, eds); err != nil {
+		return fmt.Errorf("writing Q4: %w", err)
+	}
+
+	if err := buf.Flush(); err != nil {
+		return fmt.Errorf("flushing Q4: %w", err)
+	}
+
+	if err := f.Sync(); err != nil {
+		return fmt.Errorf("syncing Q4 file: %w", err)
+	}
+
+	return nil
+}
+
+// writeQ4 writes the forth quadrant of the square to the writer. It writes the quadrant in row-major
+// order.
+func writeQ4(w io.Writer, eds *rsmt2d.ExtendedDataSquare) error {
+	half := eds.Width() / 2
+	for i := range half {
+		for j := range half {
+			shr := eds.GetCell(i+half, j+half) // TODO: Avoid copying inside GetCell
+			_, err := w.Write(shr)
+			if err != nil {
+				return fmt.Errorf("writing share: %w", err)
+			}
+		}
+	}
+	return nil
+}
+
+// OpenQ4 opens an existing Q4 file under given FS path.
+// It only reads the header with metadata. The other content
+// of the File is read lazily.
+// If file is empty, the ErrEmptyFile is returned.
+// File must be closed after usage.
+func OpenQ4(path string) (*Q4, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	hdr, err := readHeader(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Q4{
+		hdr:  hdr,
+		file: f,
+	}, nil
+}
+
+// RemoveQ4 removes Q4 file under the given FS path.
+func RemoveQ4(path string) error {
+	err := os.Remove(path)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("removing Q4 %s: %w", path, err)
+	}
+	return nil
+}
```

Comment:
Looks same as RemoveODS. Do we even need this file specific or might move it to store? Additionally it feel to belong to store, because it operates under assumption, that file is closed, which is controlled only on store level.

Reply:
It's only used to cleanup file on failed create. Doesn't hurt to have it, but can remove. Feel indifferent.